src/store/mod.rs

```diff
@@ -0,0 +1,147 @@
+use crate::node_pages_map::{Page, PAGE_SIZE};
+use io_uring::{cqueue, opcode, squeue, types, IoUring};
+use slab::Slab;
+use std::{
+    fs::{File, OpenOptions},
+    io,
+    ops::Range,
+    os::{fd::AsRawFd, unix::fs::OpenOptionsExt},
+    path::PathBuf,
+    sync::Arc,
+};
+
+#[cfg(test)]
+mod tests;
+
+const RING_CAPACITY: u32 = 128;
+
+/// The Store is an on disk array of [`crate::node_pages_map::Page`]
+pub struct Store {
+    store_file: File,
+}
+
+/// Options to open a Store
+pub struct StoreOptions {
+    /// File path of the storage file.
+    file_path: PathBuf,
+}
+
+impl Default for StoreOptions {
+    fn default() -> Self {
+        Self {
+            file_path: PathBuf::from("node_pages_store"),
+        }
+    }
+}
+
+impl Store {
+    /// Create a new Store given the StoreOptions
+    pub fn new(options: StoreOptions) -> Self {
+        let store_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(libc::O_DIRECT)
+            .open(options.file_path)
+            .expect("could not open file");
+
+        Store { store_file }
+    }
+}
+
+/// An I/O handle for a store. This owns an io_uring ring which is used to submit reads and writes.
+pub struct StoreIo {
+    store: Arc<Store>,
+    ring: IoUring,
+    pending: Slab<Pending>,
+}
+
+impl StoreIo {
+    pub fn new(store: Arc<Store>) -> Self {
+        StoreIo {
+            ring: IoUring::<squeue::Entry, cqueue::Entry>::builder()
+                .setup_sqpoll(100)
+                // .setup_iopoll
+                .build(RING_CAPACITY)
+                .expect("Error building io_uring"),
+            store,
+            pending: Slab::with_capacity(RING_CAPACITY as _),
+        }
+    }
+
+    pub fn is_submit_full(&mut self) -> bool {
+        self.ring.submission().is_full()
```

Comment:
Attaching this comment here, but it affects all the places where you touch SQE/CQE.

This is not supposed to work, you are supposed to call `sync` when interacting with the queues (I think Gab got it right, but i did not check it thoroughly). It seems that sync has to happen before and after access to it. If you look at the source code it would become evident why.

Here is also an example from holebench:

https://github.com/pepyakin/holebench/blob/c3caef592fcd5aabeae66c514e581571c6539db2/src/backend/io_uring.rs#L85

Also, if you want to limit the number of requests in flight, you need to check the length of `pending`.

Reply:
eh, I assumed that it calls `sync` internally but I can do it manually as well.

Reply:
Actually, this is correct. It does it on `Drop`. it is only necessary to sync manually if you split the queues.